File: python/tests/hof_test.py
import functools

l = ["abc", "def"]


def test_map():
    result = ["ABC", "DEF"]
    assert [x.upper() for x in l] == result
    assert list(map(lambda x: x.upper(), l)) == result


def test_filter():
    result = ["abc"]
    assert [x for x in l if "a" in x] == result
    assert list(filter(lambda x: "a" in x, l)) == result


def test_reduce():
    result = "abcdef"
    assert functools.reduce(lambda r, x: "".join([r, x]), l, "") == result
    "".join(l) == result


def test_reverse():
    result = ["def", "abc"]
    assert list(reversed(l)) == result


def test_all():
    assert all([len(x) > 0 for x in l])


def test_any():
    assert any(["a" in x for x in l])

File: elixir/apps/challenges/test/challenges/hof_test.exs
defmodule Challenges.HofTest do
  use ExUnit.Case, async: true

  defp list do
    ["abc", "def"]
  end

  test "map" do
    result = ["ABC", "DEF"]
    assert(Enum.map(list(), fn x -> String.upcase(x) end) == result)
  end

  test "filter" do
    result = ["abc"]
    assert(Enum.filter(list(), fn x -> String.contains?(x, "a") end) == result)
  end

  test "reduce" do
    result = "abcdef"
    assert(Enum.reduce(list(), "", fn x, acc -> acc <> x end) == result)
  end

  test "reverse" do
    result = ["def", "abc"]
    assert(Enum.reverse(list()) == result)
  end

  test "all" do
    assert list()
    |> Enum.map(fn x -> String.length(x) > 0 end)
    |> Enum.all?()
  end

  test "any" do
    assert list()
    |> Enum.map(fn x -> String.contains?(x, "a") end)
    |> Enum.any?()
  end
end
